torchao/prototype/float8nocompile/float8nocompile_linear.py

```diff
@@ -19,4 +36,149 @@ class Float8LinearNoCompile(torch.nn.Linear):
+        # perform hp to fp8 conversions
+        # TODO(danielvegamyhre): replace conversion with triton kernels
+        input_fp8 = self.cast_input_to_float8(input, self.is_amax_initialized)
+        weight_scale = self.get_weight_scale(self.weight)
+        weight_fp8_t = self.cast_weight_to_float8_t(
+            self.weight, self.is_amax_initialized, weight_scale
+        )
+
+        # compute fp8 matmul
+        output = manual_float8_matmul_with_args_in_float8.apply(input_fp8, weight_fp8_t)
+
+        # cast grad_output to float8_e5m2 during backward
+        # TODO(danielvegamyhre): replace with triton kernel
+        return self.cast_output_to_float8_in_bw(output)
+
+    def cast_input_to_float8(
+        self, input: torch.Tensor, is_amax_initialized: bool
+    ) -> torch.Tensor:
+        # Duplicate the autocast logic for F.linear, so that the output
+        # of our module has the right original precision
+        if torch.is_autocast_enabled():
+            # For now, hardcode to GPU's autocast dtype
+            # if we need CPU support in the future, we can add it
+            autocast_dtype = torch.get_autocast_gpu_dtype()
+            input = input.to(autocast_dtype)
+
+        # TODO(danielvegamyhre): implement this fn in scaling_utils with call to triton kernel
+        return hp_tensor_to_float8nocompile_dynamic(
+            input,
+            self.config.cast_config_input.target_dtype,
+            self.linear_mm_config,
+            gemm_input_role=GemmInputRole.INPUT,
+        )
+
+    def get_weight_scale(self, weight: torch.Tensor) -> Optional[torch.Tensor]:
+        # TODO(danielvegamyhre): replace scale calculation with triton kernel
+        if tensor_already_casted_to_fp8(weight):
+            return None
+        return tensor_to_scale(weight, self.config.cast_config_weight.target_dtype)
+
+    def cast_weight_to_float8_t(
+        self,
+        weight: torch.Tensor,
+        is_amax_initialized: bool,
+        weight_scale: Optional[torch.Tensor] = None,
+    ) -> torch.Tensor:
+        if tensor_already_casted_to_fp8(weight):
+            return weight.t()
+
+        # TODO(danielvegamyhre): replace conversion with triton kernel
+        weight_fp8 = hp_tensor_and_scale_to_float8(
+            weight,
+            weight_scale,
+            self.config.cast_config_weight.target_dtype,
+            self.linear_mm_config,
+            gemm_input_role=GemmInputRole.WEIGHT,
+        )
+        return weight_fp8.t()
+
+    def cast_output_to_float8_in_bw(self, output: torch.Tensor) -> torch.Tensor:
+        # TODO(danielvegamyhre): replace conversion with triton kernel
+        return NoopFwToFloat8BwDynamic.apply(
+            output,
+            self.linear_mm_config,
+            self.config.cast_config_grad_output.target_dtype,
+        )
+
+    @classmethod
+    def from_float(
+        cls,
+        mod,
+        config: Optional[Float8LinearConfig] = None,
```

Comment:
nit: feel free to only support default settings and not give an option to modify the config